wptrunner/executors/pytestrunner/fixtures.py

```diff
@@ -45,14 +46,66 @@ def setup(request, session):
         def test_something(setup, session):
             assert session.url == "https://example.org"
 
-    The session is closed when the test module goes out of scope by an
-    implicit call to `session.end`.
+    When the test function goes out of scope, any remaining user prompts
+    and opened windows are closed, and the current browsing context is
+    switched back to the top-level browsing context.
     """
 
     def __init__(self, client):
         self.client = client
 
-    @pytest.fixture(scope="module")
+    @pytest.fixture(scope="function")
     def session(self, request):
-        request.addfinalizer(self.client.end)
+        # finalisers are popped off a stack,
+        # making their ordering reverse
+        request.addfinalizer(self.switch_to_top_level_browsing_context)
+        request.addfinalizer(self.restore_windows)
+        request.addfinalizer(self.dismiss_user_prompts)
+
         return self.client
+
+    def dismiss_user_prompts(self):
+        """Dismisses any open user prompts in windows."""
+        current_window = self.client.window_handle
+
+        for window in self.windows():
+            self.client.window_handle = window
+            try:
+                self.client.alert.dismiss()
+            except webdriver.NoSuchAlertException:
+                pass
+
+        self.client.window_handle = current_window
+
+    def restore_windows(self):
+        """Closes superfluous windows opened by the test without ending
+        the session implicitly by closing the last window.
+        """
+        for window in self.windows():
+            self.client.window_handle = window
+            if len(self.client.window_handles) > 1:
+                self.client.close()
+
+    def switch_to_top_level_browsing_context(self):
+        """If the current browsing context selected by WebDriver is a
+        `<frame>` or an `<iframe>`, switch it back to the top-level
+        browsing context.
+        """
+        self.client.switch_frame(None)
+
+    def windows(self, exclude=None):
+        """Iterator for windows.
+        
+        :param exclude: If provided with a list of window handles,
+            these are excluded from iteration.
+
+        :returns: Iterator over window handles.
```

Comment:
Not sure why you're forcing an iterator here, why not just return a set? (Or list if you don't want to do a conversion)